### PR TITLE
Convert over to using Iterators and Yield

### PR DESCRIPTION
**Pull request by AdamSpeight2008**
First part of work extending the use of Iterators and Yield, see #27240

### CODE REVIEW

File: src/EditorFeatures/VisualBasic/Highlighting/KeywordHighlighters/AccessorDeclarationHighlighter.vb [hunk 1]
```diff
@@ -10,45 +10,38 @@ Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.KeywordHighlighting
...
-            If methodBlock Is Nothing OrElse Not TypeOf methodBlock.BlockStatement Is AccessorStatementSyntax Then
-                Return SpecializedCollections.EmptyEnumerable(Of TextSpan)()
-            End If
-
-            Dim highlights As New List(Of TextSpan)()
+        Protected Overloads Overrides Iterator Function GetHighlights(node As SyntaxNode, cancellationToken As CancellationToken) As IEnumerable(Of TextSpan)
+            If cancellationToken.IsCancellationRequested Then Return
 
+            Dim methodBlock = node.GetAncestor(Of MethodBlockBaseSyntax)()
+            If methodBlock Is Nothing OrElse Not TypeOf methodBlock.BlockStatement Is AccessorStatementSyntax Then Return
```
Comment (jasonmalinowski) [inline src/EditorFeatures/VisualBasic/Highlighting/KeywordHighlighters/AccessorDeclarationHighlighter.vb]: Roslyn code style here is we should still be putting the returns on their own line with an end-if. Just as we wouldn't write `if (blah) return;` in C#. Any chance you can do a pass and update this PR?

Comment (AdamSpeight2008) [reply src/EditorFeatures/VisualBasic/Highlighting/KeywordHighlighters/AccessorDeclarationHighlighter.vb]: @jasonmalinowski  Pass Done

File: src/EditorFeatures/VisualBasic/Highlighting/KeywordHighlightingHelpers.vb [hunk 2]
```diff
@@ -8,127 +8,112 @@ Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
...
+                End If
+
+                Dim children = node.ChildNodes.Where(Function(child) TypeOf child IsNot LambdaExpressionSyntax)
+                nodes.AddRangeAtHead(children)
+
+            End While
+        End Function
+
+        Private Iterator Function HighlightRelatedYieldStatements(Of T)(thisnode As SyntaxNode) As IEnumerable(Of TextSpan)
+            Dim nodes As New LinkedList(Of SyntaxNode)
```
Comment (AdamSpeight2008) [inline src/EditorFeatures/VisualBasic/Highlighting/KeywordHighlightingHelpers.vb]: @jasonmalinowski  I'm using a `LinkedList` to help preserve the order of evaluation, due the pre-pull request recursion.
Don't know if there is a better method.

File: src/EditorFeatures/VisualBasic/Highlighting/KeywordHighlightingHelpers.vb [hunk 3]
```diff
@@ -8,127 +8,112 @@ Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
...
-
-            Return False
+            Return (exitStatement IsNot Nothing) AndAlso (exitStatement.Kind <> expectedKind)
         End Function
 
         <Extension>
-        Friend Sub HighlightRelatedAwaits(node As SyntaxNode, highlights As List(Of TextSpan), cancellationToken As CancellationToken)
-            If TypeOf node Is AwaitExpressionSyntax Then
-                With DirectCast(node, AwaitExpressionSyntax)
-                    ' If there is already a highlight for the previous token and it is on the same line,
```
Comment (AdamSpeight2008) [inline src/EditorFeatures/VisualBasic/Highlighting/KeywordHighlightingHelpers.vb]: Maybe able to replicate this via an extension method, to preserve existing highlights,

**Comment by Neme12** [comment]
What is the benefit of this refactoring? This PR changes almost 470 lines of code with no apparent purpose. In fact it adds a few extra lines rather than reducing/simplifying anything.
image

Can you clarify why this change would be useful? Thanks

**Comment by Neme12** [comment]
If you're going to do this, please consider changing the C# equivalent of this as well rather than adding more inconsistency.

**Comment by AdamSpeight2008** [comment]
@Neme12 Previously the highlighters were creating a new list which was enumerated over, which is / was a common pattern for the lack of iterator and yield in VB. VB now supports them, so this change remove that list creation, and directly enumerates them.

**Comment by AdamSpeight2008** [comment]
@Neme12  If I do the C# implementations it would be in a separate pull request, (ie pending this pr's ok).

**Comment by Neme12** [comment]
> VB now supports them, so this change remove that list creation, and directly enumerates them.

Can you explain why that is better than creating a list, especially considering the use of recursion here?

**Comment by Neme12** [comment]
I can't speak for VB but I do know this pattern is also used in the C# code. Yield was introduced in C# 2.0 and I do not believe this code predates that.

**Comment by AdamSpeight2008** [comment]
The original pr was `+357 −527` but I had to change if statement style.  The change may use less memory and VS should be slight more responsive  *(I don't know how validate these though)*

**Comment by Neme12** [comment]
> VS should be slight more responsive

Why would creating a new enumerable and a state machine at each recursion level be more performant?

**Comment by Neme12** [comment]
> The original pr was +357 −527 but I had to change if statement style. 

So the original reduction was purely because of your style changes, not the use of yield.

**Comment by AdamSpeight2008** [comment]
@Neme12 0
> Can you explain why that is better than creating a list, especially considering the use of recursion here? 

I wouldn't be using recursion with iterators as that can make the algorithm order of `n^2`. In this pull request I am removing the recursion, both so I can linearise the algorithm and use yield. The mention of recursion was about preserving the order of evaluate of the nodes, as per the existing code (as to make it deterministic). There is little (if any recursion) use in this change.

> Why would creating a new enumerable and a state machine at each recursion level be more performant?

The ide doesn't have to wait for the full list of nodes (it lazy evaluated), before displaying them. Hence would be perceived as being more responsive / performant. I've wired in the cancellation token, so it exits earlier, so on cancellation it reacts faster. Now doesn't use exceptions on cancellation.

**Comment by AdamSpeight2008** [comment]
@dotnet-bot  retest windows_debug_vs-integration_prtest please.

**Comment by AdamSpeight2008** [comment]
@jasonmalinowski Ready for review.

**Comment by CyrusNajmabadi** [comment]
My preference here would be that these became ImmutableArray returning functions.  Or if they were passed pooled builders/lists and filled those.  That would fit with a lot more of our IDE code that is sensitive to unnecessary allocations.

**Comment by jasonmalinowski** [comment]
> The ide doesn't have to wait for the full list of nodes (it lazy evaluated), before displaying them. Hence would be perceived as being more responsive / performant.

**Comment by jasonmalinowski** [comment]
> The ide doesn't have to wait for the full list of nodes (it lazy evaluated), before displaying them. Hence would be perceived as being more responsive / performant.

Ultimately we are feeding this in to Roslyn's async tagging infrastructure; the user won't ever "see" the intermediate state of us having half-enumerated things.

**Comment by jasonmalinowski** [comment]
I suspect one reason we may have (generally) used iterator patterns for tagging is the list is *not* the final representation -- we end up putting these into tag trees, so any list you're creating is ultimately being thrown out in the end. For keyword highlighting though, I'm not sure how much that might matter; if we were regularly creating "big" lists the large lists aren't good for memory _either_.

To me I don't see a reason _not_ to take this PR, unless we concluded we didn't want to use iterators, and then somebody should put a PR on my desk removing them from C#. :smile:

File: src/EditorFeatures/VisualBasic/Highlighting/KeywordHighlightingHelpers.vb (hunk 3 above)
Comment (AdamSpeight2008) [reply src/EditorFeatures/VisualBasic/Highlighting/KeywordHighlightingHelpers.vb]: @jasonmalinowski Are you ok with the divergence from merging textspans of highlights, to separated highlights.

**Pull request closed without merging**